### PR TITLE
HA master: add firewall rule for etcd cluster.

### PR DESCRIPTION
**Pull request by jszczepkowski**
HA master: add firewall rule for accessing & securing etcd cluster.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33094)

### CODE REVIEW

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -660,7 +660,17 @@ function create-network() {
       --project "${PROJECT}" \
       --network "${NETWORK}" \
       --source-ranges "10.0.0.0/8" \
-      --allow "tcp:1-65535,udp:1-65535,icmp" &
+      --allow "tcp:1-2379,tcp:2382-65535,udp:1-65535,icmp" &
+  fi
+
+  # Create rule for accessing and securing etcd servers.
```
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: I think this should be created in `create-master` method where we also create other firewall rules specific for the master.

File: cluster/gce/util.sh [hunk 2]
```diff
@@ -660,7 +660,17 @@ function create-network() {
...
+  if ! gcloud compute firewall-rules --project "${PROJECT}" describe "${MASTER_NAME}-etcd" &>/dev/null; then
+    gcloud compute firewall-rules create "${MASTER_NAME}-etcd" \
+      --project "${PROJECT}" \
+      --network "${NETWORK}" \
+      --source-tags "${MASTER_NAME}" \
```
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: s/MASTER_NAME/MASTER_TAG

File: cluster/gce/util.sh [hunk 3]
```diff
@@ -660,7 +660,17 @@ function create-network() {
...
+      --allow "tcp:2380,tcp:2381" \
+      --target-tags "${MASTER_NAME}" &
```
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: ditto

File: cluster/gce/util.sh [hunk 4]
```diff
@@ -660,7 +660,17 @@ function create-network() {
       --project "${PROJECT}" \
       --network "${NETWORK}" \
       --source-ranges "10.0.0.0/8" \
-      --allow "tcp:1-65535,udp:1-65535,icmp" &
+      --allow "tcp:1-2379,tcp:2382-65535,udp:1-65535,icmp" &
```
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: This rule means that accessing this port on any machine, including regular nodes, is not allowed. I think we should only limit this to the master. WDYT?

**Review by fgrzadkowski — CHANGES_REQUESTED** [review]

File: cluster/gce/util.sh (hunk 2 above)
Comment (jszczepkowski) [inline cluster/gce/util.sh]: Done

File: cluster/gce/util.sh (hunk 3 above)
Comment (jszczepkowski) [inline cluster/gce/util.sh]: Done

File: cluster/gce/util.sh (hunk 1 above)
Comment (jszczepkowski) [inline cluster/gce/util.sh]: Done

File: cluster/gce/util.sh (hunk 4 above)
Comment (jszczepkowski) [inline cluster/gce/util.sh]: I agree, but I'm not sure how to implement it. Firewall rules on GCE allow only to allow certain tags. So, we can allow nodes as they have NODE_TAG, but we cannot allow everything but MASTER_TAG.

Comment (jszczepkowski) [inline cluster/gce/util.sh]: I agree, but I'm not sure how to implement it. Firewall rules on GCE allow only to allow certain tags. So, we can allow nodes as they have NODE_TAG, but we cannot allow everything but MASTER_TAG.

**Comment by jszczepkowski** [comment]
@k8s-bot gci gce e2e test this
issue: #33725

File: cluster/gce/util.sh [hunk 5]
```diff
@@ -660,7 +660,7 @@ function create-network() {
...
```
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: I guess you'd have to have two separate rules:
* for master node something which explicitly lists ports that we want to expose
* for minion nodes something that will open all ports

Would it work?

Comment (jszczepkowski) [inline cluster/gce/util.sh]: There is a slight semantic different: before we were allowing all the nodes, now only minions and masters.

However, this sounds reasonable. Done

File: cluster/gce/util.sh [hunk 6]
```diff
@@ -655,12 +655,22 @@ function create-network() {
...
 
-  if ! gcloud compute firewall-rules --project "${PROJECT}" describe "${NETWORK}-default-internal" &>/dev/null; then
-    gcloud compute firewall-rules create "${NETWORK}-default-internal" \
+  if ! gcloud compute firewall-rules --project "${PROJECT}" describe "${NETWORK}-default-internal-master" &>/dev/null; then
+    gcloud compute firewall-rules create "${NETWORK}-default-internal-master" \
       --project "${PROJECT}" \
       --network "${NETWORK}" \
       --source-ranges "10.0.0.0/8" \
-      --allow "tcp:1-65535,udp:1-65535,icmp" &
+      --allow "tcp:1-2379,tcp:2382-65535,udp:1-65535,icmp" \
```
Comment (fgrzadkowski) [inline cluster/gce/util.sh]: As explained in my previous email I think we should only expose specific ports. At the very least please leave a TODO to check which ports are actually needed on master.

Comment (fgrzadkowski) [inline cluster/gce/util.sh]: Also I think it'd be useful to explain in a comment why do you exclude those two ports.

**Review by fgrzadkowski — CHANGES_REQUESTED** [review]

**Comment by fgrzadkowski** [comment]
LGTM. One minor change. Please make sure this doesn't break e2e tests.

**Comment by ixdy** [comment]
This PR basically blew up all of our PR testing. The `-internal` firewall was never cleaned up (I'm not sure why), and now we have `-internal`, `internal-ssh`, and `-internal-node` rules, which are completely eating up all of our quota.

It also seems like the `-master-etcd` rule isn't being cleaned up for some reason, and I'm still trying to figure out why.

cc @kubernetes/test-infra-maintainers

File: cluster/gce/util.sh [hunk 7]
```diff
@@ -1261,6 +1281,13 @@ function kube-down() {
         --quiet \
         "${NODE_TAG}-all"
     fi
+    # Delete firewall rule for etcd servers.
+    if ! gcloud compute firewall-rules --project "${PROJECT}" describe "${MASTER_NAME}-etcd" &>/dev/null; then
```
Comment (ixdy) [inline cluster/gce/util.sh]: oh, this should not have the `!` in front.

Comment (ixdy) [inline cluster/gce/util.sh]: #33963

**Pull request merged**